File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/tents/domain/TentsGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.tents.domain;

import com.zwstudio.logicpuzzlesandroid.common.domain.AllowedObjectState;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import fj.F;
import fj.F0;

public class TentsGameState extends CellsGameState<TentsGame, TentsGameMove, TentsGameState> {
    public TentsObject[] objArray;
    public HintState[] row2state;
    public HintState[] col2state;

    public TentsGameState(TentsGame game) {
        super(game);
        objArray = new TentsObject[rows() * cols()];
        for (int i = 0; i < objArray.length; i++)
            objArray[i] = new TentsEmptyObject();
        for (Position p : game.pos2tree)
            set(p, new TentsTreeObject());
        row2state = new HintState[rows()];
        col2state = new HintState[cols()];
        updateIsSolved();
    }

    public TentsObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public TentsObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, TentsObject obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, TentsObject obj) {
        set(p.row, p.col, obj);
    }

    public boolean setObject(TentsGameMove move) {
        if (!isValid(move.p) || get(move.p) == move.obj) return false;
        set(move.p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(TentsGameMove move) {
        MarkerOptions markerOption = MarkerOptions.values()[game.gdi.getMarkerOption()];
        F<TentsObject, TentsObject> f = obj -> {
            if (obj instanceof TentsEmptyObject)
                return markerOption == MarkerOptions.MarkerFirst ?
                        new TentsMarkerObject() : new TentsTentObject();
            if (obj instanceof TentsTentObject)
                return markerOption == MarkerOptions.MarkerLast ?
                        new TentsMarkerObject() : new TentsEmptyObject();
            if (obj instanceof TentsMarkerObject)
                return markerOption == MarkerOptions.MarkerFirst ?
                        new TentsTentObject() : new TentsEmptyObject();
            return obj;
        };
        Position p = move.p;
        if (!isValid(p)) return false;
        move.obj = f.f(get(p));
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 1/Tents

        Summary
        Each camper wants to put his Tent under the shade of a Tree. But he also
        wants his privacy!

        Description
        1. The board represents a camping field with many Trees. Campers want to set
           their Tent in the shade, horizontally or vertically adjacent to a Tree(not
           diagonally).
        2. At the same time they need their privacy, so a Tent can't have any other
           Tents near them, not even diagonally.
        3. The numbers on the borders tell you how many Tents there are in that row
           or column.
        4. Finally, each Tree has at least one Tent touching it, horizontally or
           vertically.
    */
    private void updateIsSolved() {
        boolean allowedObjectsOnly = game.gdi.isAllowedObjectsOnly();
        isSolved = true;
        for (int r = 0; r < rows(); r++) {
            int n1 = 0, n2 = game.row2hint[r];
            for (int c = 0; c < cols(); c++)
                if (get(r, c) instanceof TentsTentObject)
                    n1++;
            // 3. The numbers on the borders tell you how many Tents there are in that row.
            row2state[r] = n1 < n2 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            if (n1 != n2) isSolved = false;
        }
        for (int c = 0; c < cols(); c++) {
            int n1 = 0, n2 = game.col2hint[c];
            for (int r = 0; r < rows(); r++)
                if (get(r, c) instanceof TentsTentObject)
                    n1++;
            // 3. The numbers on the borders tell you how many Tents there are in that column.
            col2state[c] = n1 < n2 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            if (n1 != n2) isSolved = false;
        }
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                TentsObject o = get(r, c);
                if (o instanceof TentsForbiddenObject)
                    set(r, c, new TentsEmptyObject());
            }
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                TentsObject o = get(r, c);
                F0<Boolean> hasTree = () -> {
                    for (Position os : TentsGame.offset) {
                        Position p2 = p.add(os);
                        if (isValid(p2) && get(p2) instanceof TentsTreeObject)
                            return true;
                    }
                    return false;
                };
                F<Boolean, Boolean> hasTent = isTree -> {
                    for (Position os : isTree ? TentsGame.offset : TentsGame.offset2) {
                        Position p2 = p.add(os);
                        if (isValid(p2) && get(p2) instanceof TentsTentObject)
                            return true;
                    }
                    return false;
                };
                if (o instanceof TentsTentObject) {
                    // 1. The board represents a camping field with many Trees. Campers want to set
                    // their Tent in the shade, horizontally or vertically adjacent to a Tree(not
                    // diagonally).
                    // 2. At the same time they need their privacy, so a Tent can't have any other
                    // Tents near them, not even diagonally.
                    TentsTentObject o2 = (TentsTentObject)o;
                    AllowedObjectState s = hasTree.f() && !hasTent.f(false) ? AllowedObjectState.Normal : AllowedObjectState.Error;
                    o2.state = s;
                    if (s == AllowedObjectState.Error) isSolved = false;
                } else if (o instanceof TentsTreeObject) {
                    // 4. Finally, each Tree has at least one Tent touching it, horizontally or
                    // vertically.
                    TentsTreeObject o2 = (TentsTreeObject)o;
                    AllowedObjectState s = hasTent.f(true) ? AllowedObjectState.Normal : AllowedObjectState.Error;
                    o2.state = s;
                    if (s == AllowedObjectState.Error) isSolved = false;
                } else if ((o instanceof TentsEmptyObject || o instanceof TentsMarkerObject) && allowedObjectsOnly &&
                        (col2state[c] != HintState.Normal || row2state[r] != HintState.Normal || !hasTree.f() || hasTent.f(false)))
                    set(r, c, new TentsForbiddenObject());
            }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/tents/domain/TentsGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.tents.domainimport

import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class TentsGameState(game: TentsGame) : CellsGameState<TentsGame, TentsGameMove, TentsGameState>(game) {
    var objArray: Array<TentsObject?>
    var row2state: Array<HintState?>
    var col2state: Array<HintState?>
    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]

    operator fun get(p: Position) = get(p!!.row, p.col)

    operator fun set(row: Int, col: Int, obj: TentsObject?) {
        objArray[row * cols() + col] = obj
    }

    operator fun set(p: Position, obj: TentsObject?) {
        set(p!!.row, p.col, obj)
    }

    fun setObject(move: TentsGameMove): Boolean {
        if (!isValid(move.p) || get(move.p) === move.obj) return false
        set(move.p, move.obj)
        updateIsSolved()
        return true
    }

    fun switchObject(move: TentsGameMove): Boolean {
        val markerOption: MarkerOptions = MarkerOptions.values().get(game.gdi.getMarkerOption())
        val f: F<TentsObject, TentsObject> = label@ F<TentsObject, TentsObject> { obj: TentsObject? ->
            if (obj is TentsEmptyObject) return@label if (markerOption == MarkerOptions.MarkerFirst) TentsMarkerObject() else TentsTentObject()
            if (obj is TentsTentObject) return@label if (markerOption == MarkerOptions.MarkerLast) TentsMarkerObject() else TentsEmptyObject()
            if (obj is TentsMarkerObject) return@label if (markerOption == MarkerOptions.MarkerFirst) TentsTentObject() else TentsEmptyObject()
            obj
        }
        val p: Position = move.p
        if (!isValid(p)) return false
        move.obj = f.f(get(p))
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 1/Tents

        Summary
        Each camper wants to put his Tent under the shade of a Tree. But he also
        wants his privacy!

        Description
        1. The board represents a camping field with many Trees. Campers want to set
           their Tent in the shade, horizontally or vertically adjacent to a Tree(not
           diagonally).
        2. At the same time they need their privacy, so a Tent can't have any other
           Tents near them, not even diagonally.
        3. The numbers on the borders tell you how many Tents there are in that row
           or column.
        4. Finally, each Tree has at least one Tent touching it, horizontally or
           vertically.
    */
    private fun updateIsSolved() {
        val allowedObjectsOnly: Boolean = game.gdi.isAllowedObjectsOnly()
        isSolved = true
        for (r in 0 until rows()) {
            var n1 = 0
            val n2: Int = game.row2hint.get(r)
            for (c in 0 until cols()) if (get(r, c) is TentsTentObject) n1++
            // 3. The numbers on the borders tell you how many Tents there are in that row.
            row2state[r] = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            if (n1 != n2) isSolved = false
        }
        for (c in 0 until cols()) {
            var n1 = 0
            val n2: Int = game.col2hint.get(c)
            for (r in 0 until rows()) if (get(r, c) is TentsTentObject) n1++
            // 3. The numbers on the borders tell you how many Tents there are in that column.
            col2state[c] = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            if (n1 != n2) isSolved = false
        }
        for (r in 0 until rows()) for (c in 0 until cols()) {
            val o: TentsObject? = get(r, c)
            if (o is TentsForbiddenObject) set(r, c, TentsEmptyObject())
        }
        for (r in 0 until rows()) for (c in 0 until cols()) {
            val p = Position(r, c)
            val o: TentsObject? = get(r, c)
            val hasTree: F0<Boolean> = label@ F0<Boolean> {
                for (os in TentsGame.offset) {
                    val p2 = p.add(os)
                    if (isValid(p2) && get(p2) is TentsTreeObject) return@label true
                }
                false
            }
            val hasTent: F<Boolean, Boolean> = label@ F<Boolean, Boolean> { isTree: Boolean ->
                for (os in if (isTree) TentsGame.offset else TentsGame.offset2) {
                    val p2 = p.add(os)
                    if (isValid(p2) && get(p2) is TentsTentObject) return@label true
                }
                false
            }
            if (o is TentsTentObject) {
                // 1. The board represents a camping field with many Trees. Campers want to set
                // their Tent in the shade, horizontally or vertically adjacent to a Tree(not
                // diagonally).
                // 2. At the same time they need their privacy, so a Tent can't have any other
                // Tents near them, not even diagonally.
                val s: AllowedObjectState = if (hasTree.f() && !hasTent.f(false)) AllowedObjectState.Normal else AllowedObjectState.Error
                o.state = s
                if (s == AllowedObjectState.Error) isSolved = false
            } else if (o is TentsTreeObject) {
                // 4. Finally, each Tree has at least one Tent touching it, horizontally or
                // vertically.
                val s: AllowedObjectState = if (hasTent.f(true)) AllowedObjectState.Normal else AllowedObjectState.Error
                o.state = s
                if (s == AllowedObjectState.Error) isSolved = false
            } else if ((o is TentsEmptyObject || o is TentsMarkerObject) && allowedObjectsOnly &&
                (col2state[c] != HintState.Normal || row2state[r] != HintState.Normal || !hasTree.f() || hasTent.f(false))) set(r, c, TentsForbiddenObject())
        }
    }

    init {
        objArray = arrayOfNulls<TentsObject>(rows() * cols())
        for (i in objArray.indices) objArray[i] = TentsEmptyObject()
        for (p in game.pos2tree) set(p, TentsTreeObject())
        row2state = arrayOfNulls<HintState>(rows())
        col2state = arrayOfNulls<HintState>(cols())
        updateIsSolved()
    }
}